File: vulnerable-demo-multilang/vuln-app.pl
#!/usr/bin/perl
# vuln-app.pl - a deliberately vulnerable demo Perl CGI script used ONLY to test
# VulnHunter's multi-language scanner coverage. DO NOT deploy this anywhere. It
# contains intentional security flaws for demonstration purposes only.
#
# Planted vulnerabilities (for scoring / demo reference):
#   1. Command injection via backticks with interpolated var -> CWE-78
#   2. eval() on untrusted input                              -> CWE-95
#   3. Hardcoded credential                                    -> CWE-798

use strict;
use warnings;
use CGI qw(:standard);

# VULN 3: Hardcoded credential (CWE-798) - should come from an environment variable
# or a secrets manager, not be committed in source.
my $DB_PASSWORD = "SuperSecretP@ss123";
my $DB_USER     = "vulnshop_admin";

my $cgi = CGI->new;

# VULN 1: Command injection (CWE-78) - the user-controlled "host" param is interpolated
# directly into a backtick shell command, letting an attacker inject shell metacharacters.
sub ping_host {
    my $host = $cgi->param('host') || '127.0.0.1';
    my $output = `ping -c 1 $host`;
    return $output;
}

# VULN 2: eval() on untrusted input (CWE-95) - a string built from user input is passed
# straight to Perl's string eval, allowing arbitrary code execution.
sub calc {
    my $expr = $cgi->param('expr') || '0';
    my $result = eval "$expr";
    return $result;
}

print $cgi->header('text/plain');

if (defined $cgi->param('host')) {
    print ping_host();
}

if (defined $cgi->param('expr')) {
    print calc();
}

File: vulnerable-demo-multilang/vuln-app.js
/*
 * vuln-app.js - a deliberately vulnerable demo Express app used ONLY to test
 * VulnHunter's multi-language scanner coverage. DO NOT deploy this anywhere. It
 * contains intentional security flaws for demonstration purposes only.
 *
 * Planted vulnerabilities (for scoring / demo reference):
 *   1. Command injection via child_process.exec       -> CWE-78
 *   2. Reflected XSS via unsanitized template string   -> CWE-79
 *   3. Hardcoded API key                                -> CWE-798
 */

const express = require("express");
const { exec } = require("child_process");

const app = express();

// VULN 3: Hardcoded secret (CWE-798) - should come from process.env / a secrets manager
const STRIPE_API_KEY = "sk_live_DEMO_FAKE_NOT_A_REAL_KEY_9876543210";

app.get("/ping", (req, res) => {
  // VULN 1: Command injection (CWE-78) - user-controlled host is interpolated
  // straight into a shell command string instead of using execFile with an args array.
  const host = req.query.host || "127.0.0.1";
  exec(`ping -c 1 ${host}`, (err, stdout, stderr) => {
    if (err) {
      res.status(500).send("ping failed");
      return;
    }
    res.send(stdout);
  });
});

app.get("/greet", (req, res) => {
  // VULN 2: Reflected XSS (CWE-79) - user-controlled "name" is concatenated directly
  // into an HTML response template string with no escaping, equivalent to assigning
  // untrusted data straight into innerHTML on the client.
  const name = req.query.name || "guest";
  const page = `<html><body><h1>Welcome back, ${name}!</h1></body></html>`;
  res.set("Content-Type", "text/html");
  res.send(page);
});

app.get("/charge", (req, res) => {
  // Demonstrates the hardcoded key actually being used.
  const amount = req.query.amount;
  res.json({ status: "charged", amount, key_used: STRIPE_API_KEY.slice(0, 10) + "..." });
});

app.listen(3000, () => {
  console.log("vuln-app listening on port 3000");
});

module.exports = app;
